### DISCUSSION TITLE
Gnosis Safe Signature Support

### DISCUSSION THREAD

**Discussion by shrugs** [wevm/wagmi#90] [Ideas]
Any ideas for supporting on-chain signatures via EIP-1271 as part of the `useSignMessage` hook? This would let contracts like Gnosis Safe sign a message for authorization.

[EIP-1271: Standard Signature Validation Method for Contracts](https://eips.ethereum.org/EIPS/eip-1271)

**Comment by sammdec** [comment]
Does `useSignMessage` not generate a valid signature to be used for EIP-1271?

**Reply by shrugs** [reply]
it does produce a valid message signature request (which a contract like gnosis proposes as a transaction) but the default behavior of `ethers.signMessage` doesn't actually wait for that signature to become valid or validate it against the chain in any way.

for example, here's a custom `useLogin` hook i had to write to get smart contract signatures to work with my app:

```ts
import { ethers } from 'ethers';
import pRetry from 'p-retry';
import { useCallback, useState } from 'react';
import { SignatureType, SiweMessage } from 'siwe';
import { useAccount, useNetwork, useProvider, useSigner } from 'wagmi';

import { NonceResponse } from '@/pages/api/nonce';

import { AVERAGE_BLOCK_TIME_IN_SECONDS } from './constants';
import { verifyEIP1271Signature } from './eip1271';
import fetcher, { FetchError } from './fetcher';
import useUser from './useUser';

export enum LoginState {
  Idle,
  Signing,
  Confirming,
  Authenticating,
}

export function useLogin() {
  const provider = useProvider();
  const [{ data: signer }] = useSigner();
  const [
    {
      data: { chain },
    },
  ] = useNetwork();
  const [{ data }] = useAccount();
  const address = data?.address;

  const { mutate: mutateUser } = useUser();
  const [state, setState] = useState(LoginState.Idle);
  const [error, setError] = useState<Error>();

  const login = useCallback(async () => {
    setState(LoginState.Signing);
    setError(undefined);

    try {
      if (!provider) throw new Error(`Cannnot sign in without provider.`);
      if (!address) throw new Error('Address not found or provided.');

      const [{ nonce }, code] = await Promise.all([
        fetcher<NonceResponse>(`/api/nonce`),
        provider.getCode(address),
      ]);

      const isContractAccount = code !== '0x';

      const message = new SiweMessage({
        domain: document.location.host,
        address,
        chainId: chain.id.toString(),
        uri: document.location.origin,
        version: '1',
        statement: 'Pleasr Auction House',
        type: SignatureType.PERSONAL_SIGNATURE,
        nonce,
      });

      const signature = await signer.signMessage(message.signMessage());
      // TODO: curious bug related to wallet connect hanging on this line even after signing ^
      // seems to occur after disconnect and reload, request goes through but promise never resolves
      // i fucking hate walletconnect.
      message.signature = signature;

      // if contract and not already valid, wait for on-chain confirmation
      if (isContractAccount) {
        setState(LoginState.Confirming);
        const signedData = ethers.utils.hashMessage(message.signMessage());
        // wait for tx to go through
        await pRetry(
          async () => {
            console.log('Checking for resolved contract signature...');
            const valid = await verifyEIP1271Signature(
              provider,
              message.address,
              signedData,
              message.signature,
            );

            if (!valid) throw new Error('Not yet valid contract signature...');
          },
          {
            forever: true,
            minTimeout: AVERAGE_BLOCK_TIME_IN_SECONDS * 1000,
            maxTimeout: 5 * AVERAGE_BLOCK_TIME_IN_SECONDS * 1000,
          },
        );
      }

      setState(LoginState.Authenticating);

      await mutateUser(
        fetcher(`/api/login`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ message }),
        }),
        { revalidate: false },
      );
    } catch (error) {
      if (error instanceof FetchError) {
        setError(new Error(error.data.message));
      } else if (error instanceof Error) {
        setError(error);
      } else {
        throw error;
      }
    } finally {
      setState(LoginState.Idle);
    }
  }, [address, chain, mutateUser, provider, signer]);

  return { login, state, error };
}

```

**Reply by sammdec** [reply]
Ah that’s interesting, that might be why I couldn’t get out auth to work for our app when using gnosis.

**Reply by shrugs** [reply]
here's the source of the eip1271 file for future reference/impl:


```ts
import { ethers } from 'ethers';

const MAGIC_VALUE = '0x1626ba7e';
const EIP_1271_ABI = [
  `function isValidSignature(bytes32 _hash, bytes memory _signature) public view returns (bytes4 magicValue)`,
];

export async function verifyEIP1271Signature(
  provider: ethers.providers.BaseProvider,
  address: string,
  data: string,
  signature: string,
): Promise<boolean> {
  const contract = new ethers.Contract(address, EIP_1271_ABI, provider);

  try {
    const returnValue = await contract['isValidSignature'](data, signature);
    return returnValue.toLowerCase() === MAGIC_VALUE.toLowerCase();
  } catch (error) {
    console.error(error);
    return false;
  }
}

```

**Reply by shrugs** [reply]
by default gnosis (and i expect any other contract wallets that require an on-chain signature) will return `0x` as the response to `ethers.signMessage` after the transaction has been proposed and that promise will resolve. so without special knowledge about waiting for the signature to actually become valid, your app just gets an empty signature and bails

**Reply by shrugs** [reply]
fwiw i've tested this with Argent but their model doesn't require an on-chain signature so the value returned by `ethers.signMessage` is valid but should still be verified with eip1271 to allow someone to log in as their argent smart wallet

**Reply by sammdec** [reply]
Right I see so we would have to listen for tx to become successful get the result which should be the signature and then use that in the eip-1271 `isValidSignature` Contract call?

**Reply by tmm** [reply]
@shrugs a lot of great knowledge shared here! would be cool to package this up for folks to benefit from.

Any interest in kicking off a PR? Otherwise, I can add to my backlog.

**Reply by shrugs** [reply]
> Right I see so we would have to listen for tx to become successful get the result which should be the signature and then use that in the eip-1271 `isValidSignature` Contract call?

mostly — after the tx is confirmed (which my code doesn't actually check because we don't get the tx hash back from `signMessage`) the hash will be valid on-chain, which is what we're actually checking within that `pRetry` block. the code could be made more efficient if we knew the txhash but there's no easy way to get that afaik, though i haven't looked hard.

> Any interest in kicking off a PR? Otherwise, I can add to my backlog.

i'm heads down right now and can't stand up a PR but please do borrow my code as necessary!

**Reply by sammdec** [reply]
This is incredibly helpful to know, great insight thanks!

**Reply by tmm** [reply]
Another interesting approach.

https://gallery.mirror.xyz/GzHNX1UvpNmBDQjQ0pQnE_wsdxoS0_yWocRGOssH4Xw

Could be cool to make a guide for wagmi too.

**Comment by shrugs** [comment]
ah listening to the event is a good improvement. but they should use ethers hashMessage instead of writing it all out inline.

**Comment by Ivshti** [comment]
You can use this library: https://github.com/AmbireTech/signature-validator/ to support EIP1271 without any additional code. This library supports all types of signatures, including smart contract wallets (EIP 1271) out of the box
